src/coordinator/storage/local/cluster.go

```diff
@@ -0,0 +1,293 @@
+				"retention=%s, resolution=%s",
+				key.Retention.String(), key.Resolution.String())
+		}
+
+		aggregatedNamespaces[key] = namespace
+	}
+
+	return &clusters{
+		namespaces:            namespaces,
+		unaggregatedNamespace: unaggregatedNamespace,
+		aggregatedNamespaces:  aggregatedNamespaces,
+	}, nil
+}
+
+func (c *clusters) ClusterNamespaces() []ClusterNamespace {
+	return c.namespaces
+}
+
+func (c *clusters) UnaggregatedClusterNamespace() ClusterNamespace {
+	return c.unaggregatedNamespace
+}
+
+func (c *clusters) AggregatedClusterNamespace(
+	attrs RetentionResolution,
+) (ClusterNamespace, bool) {
+	namespace, ok := c.aggregatedNamespaces[attrs]
+	return namespace, ok
+}
+
+func (c *clusters) Close() error {
+	var (
+		wg             sync.WaitGroup
+		syncMultiErrs  syncMultiErrs
+		uniqueSessions []client.Session
+	)
+	// Collect unique sessions, some namespaces may share same
+	// client session (same cluster)
+	uniqueSessions = append(uniqueSessions, c.unaggregatedNamespace.Session())
+	for _, namespace := range c.aggregatedNamespaces {
+		unique := true
+		for _, session := range uniqueSessions {
+			if namespace.Session() == session {
+				unique = false
+				break
+			}
+		}
+		if unique {
+			uniqueSessions = append(uniqueSessions, namespace.Session())
+		}
+	}
+
+	for _, session := range uniqueSessions {
+		session := session // Capture for lambda
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := session.Close()
+			syncMultiErrs.add(err)
+		}()
+	}
+
+	wg.Wait()
+
+	return syncMultiErrs.finalError()
+}
+
+type clusterNamespace struct {
+	namespaceID ident.ID
+	attributes  storage.Attributes
+	session     client.Session
+}
+
+func newUnaggregatedClusterNamespace(
+	def UnaggregatedClusterNamespaceDefinition,
+) (ClusterNamespace, error) {
+	if err := def.Validate(); err != nil {
+		return nil, err
+	}
+	return &clusterNamespace{
+		namespaceID: def.NamespaceID,
+		attributes: storage.Attributes{
+			MetricsType: storage.UnaggregatedMetricsType,
+			Retention:   def.Retention,
+		},
+		session: def.Session,
+	}, nil
+}
+
+func newAggregatedClusterNamespace(
+	def AggregatedClusterNamespaceDefinition,
+) (ClusterNamespace, error) {
+	if err := def.Validate(); err != nil {
+		return nil, err
+	}
+	return &clusterNamespace{
+		namespaceID: def.NamespaceID,
+		attributes: storage.Attributes{
+			MetricsType: storage.AggregatedMetricsType,
+			Retention:   def.Retention,
+			Resolution:  def.Resolution,
+		},
+		session: def.Session,
+	}, nil
+}
+
+func (n *clusterNamespace) NamespaceID() ident.ID {
+	return n.namespaceID
+}
+
+func (n *clusterNamespace) Attributes() storage.Attributes {
+	return n.attributes
+}
+
+func (n *clusterNamespace) Session() client.Session {
+	return n.session
+}
+
+type syncMultiErrs struct {
+	sync.Mutex
+	multiErr xerrors.MultiError
+}
+
+func (errs *syncMultiErrs) add(err error) {
+	errs.Lock()
+	errs.multiErr = errs.multiErr.Add(err)
+	errs.Unlock()
```

Comment:
Yeah no big reason, if it was in the hot path I'd always avoid defer since there is some overhead there.  Since I was returning a value in the other one it is slightly easier to write, the defer way.  I'll make these both avoid defer though, just in case this does get used by someone else in a hot path.